Revert the following commits from Bug: https://trac.sagemath.org/ticket/23177
because we don't have ipywidgets 7 yet.

From e1c1fa74803213b4d596a83edd198d66ff02211b Mon Sep 17 00:00:00 2001
From: Jeroen Demeyer <jdemeyer@cage.ugent.be>
Date: Thu, 8 Jun 2017 13:59:03 +0200
Subject: Upgrade to ipywidgets 7

---
 build/pkgs/ipywidgets/checksums.ini               |   6 +-
 build/pkgs/ipywidgets/package-version.txt         |   2 +-
 build/pkgs/ipywidgets/patches/widget_repr.patch   |  45 ------
 build/pkgs/widgetsnbextension/checksums.ini       |   6 +-
 build/pkgs/widgetsnbextension/package-version.txt |   2 +-
 src/sage/interacts/test_jupyter.rst               | 171 +++++++++++-----------
 src/sage/repl/ipython_kernel/interact.py          |  34 ++---
 src/sage/repl/ipython_kernel/widgets.py           |  20 +--
 src/sage/repl/ipython_kernel/widgets_sagenb.py    |  68 ++++-----
 9 files changed, 155 insertions(+), 199 deletions(-)
 delete mode 100644 build/pkgs/ipywidgets/patches/widget_repr.patch

From e64c034e0ae46dc08ff444bb8a6b85430d2b1c26 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Fran=C3=A7ois=20Bissey?= <frp.bissey@gmail.com>
Date: Thu, 15 Feb 2018 13:27:26 +1300
Subject: Fix doctest in cpython/debug.pyx

---
 src/sage/cpython/debug.pyx | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/sage/src/sage/interacts/test_jupyter.rst
+++ b/sage/src/sage/interacts/test_jupyter.rst
@@ -6,7 +6,7 @@
 
 We need to setup a proper test environment for widgets::
 
-    sage: from ipywidgets.widgets.tests.utils import setup_test_comm
+    sage: from ipywidgets.widgets.tests import setup_test_comm
     sage: setup_test_comm()
 
 Make sure that we use the Jupyter interacts::
@@ -39,12 +39,12 @@
 
     sage: test(interacts.algebra.polar_prime_spiral)  # long time
     Interactive function <function polar_prime_spiral at ...> with 6 widgets
-      interval: IntRangeSlider(value=(1, 1000), description=u'range', max=4000, min=1, step=10)
+      interval: IntRangeSlider(value=(1, 1000), min=1, max=4000, step=10, description=u'range')
       show_factors: Checkbox(value=True, description=u'show_factors')
       highlight_primes: Checkbox(value=True, description=u'highlight_primes')
       show_curves: Checkbox(value=True, description=u'show_curves')
-      n: IntSlider(value=89, description=u'number $n$', max=200, min=1)
-      dpi: IntSlider(value=100, description=u'dpi', max=300, min=10, step=10)
+      n: IntSlider(value=89, min=1, max=200, step=1, description=u'number $n$')
+      dpi: IntSlider(value=100, min=10, max=300, step=10, description=u'dpi')
     <h2>Polar Prime Spiral</h2>           <div style="white-space: normal;">          For more information about the factors in the spiral, visit           <a href="http://www.dcs.gla.ac.uk/~jhw/spirals/index.html" target="_blank">          Number Spirals by John Williamson</a>.</div>
     <script type="math/tex">n = 89</script>
     Pink Curve:  <script type="math/tex">n^2 + 8</script>
@@ -53,27 +53,27 @@
     sage: test(interacts.calculus.taylor_polynomial)
     Interactive function <function taylor_polynomial at ...> with 3 widgets
       title: HTMLText(value=u'<h2>Taylor polynomial</h2>')
-      f: EvalText(value=u'e^(-x)*sin(x)', description=u'$f(x)=$', layout=Layout(max_width=u'81em'))
-      order: SelectionSlider(description=u'order', options=(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12), value=1)
+      f: EvalText(value=u'e^(-x)*sin(x)', description=u'$f(x)=$')
+      order: SelectionSlider(value=1, options=[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12], description=u'order')
     <script type="math/tex">f(x)\;=\;e^{\left(-x\right)} \sin\left(x\right)</script>
     <script type="math/tex">\hat{f}(x;0)\;=\;x+\mathcal{O}(x^{2})</script>
 
     sage: test(interacts.calculus.definite_integral)
     Interactive function <function definite_integral at ...> with 6 widgets
       title: HTMLText(value=u'<h2>Definite integral</h2>')
-      f: EvalText(value=u'3*x', description=u'$f(x)=$', layout=Layout(max_width=u'81em'))
-      g: EvalText(value=u'x^2', description=u'$g(x)=$', layout=Layout(max_width=u'81em'))
-      interval: IntRangeSlider(value=(0, 3), description=u'Interval', max=10, min=-10)
-      x_range: IntRangeSlider(value=(0, 3), description=u'plot range (x)', max=10, min=-10)
-      selection: Dropdown(description=u'Select', index=2, options=('f', 'g', 'f and g', 'f - g'), value='f and g')
+      f: EvalText(value=u'3*x', description=u'$f(x)=$')
+      g: EvalText(value=u'x^2', description=u'$g(x)=$')
+      interval: IntRangeSlider(value=(0, 3), min=-10, max=10, step=1, description=u'Interval')
+      x_range: IntRangeSlider(value=(0, 3), min=-10, max=10, step=1, description=u'plot range (x)')
+      selection: Dropdown(value='f and g', options=['f', 'g', 'f and g', 'f - g'], description=u'Select')
     <script type="math/tex">\int_{0.00}^{3.00}(\color{Blue}{f(x)})\,\mathrm{d}x=\int_{0.00}^{3.00}(3 \, x)\,\mathrm{d}x=13.50</script><br/><script type="math/tex">\int_{0.00}^{3.00}(\color{Green}{g(x)})\,\mathrm{d}x=\int_{0.00}^{3.00}(x^{2})\,\mathrm{d}x=9.00</script>
 
     sage: test(interacts.calculus.function_derivative)
     Interactive function <function function_derivative at ...> with 4 widgets
       title: HTMLText(value=u'<h2>Derivative grapher</h2>')
-      function: EvalText(value=u'x^5-3*x^3+1', description=u'Function:', layout=Layout(max_width=u'81em'))
-      x_range: FloatRangeSlider(value=(-2.0, 2.0), description=u'Range (x)', max=15.0, min=-15.0)
-      y_range: FloatRangeSlider(value=(-8.0, 6.0), description=u'Range (y)', max=15.0, min=-15.0)
+      function: EvalText(value=u'x^5-3*x^3+1', description=u'Function:')
+      x_range: FloatRangeSlider(value=(-2.0, 2.0), min=-15.0, max=15.0, step=0.1, description=u'Range (x)')
+      y_range: FloatRangeSlider(value=(-8.0, 6.0), min=-15.0, max=15.0, step=0.1, description=u'Range (y)')
     <center><script type="math/tex">\color{Blue}{f(x) = x^{5} - 3 \, x^{3} + 1}</script></center>
     <center><script type="math/tex">\color{Green}{f'(x) = 5 \, x^{4} - 9 \, x^{2}}</script></center>
     <center><script type="math/tex">\color{Red}{f''(x) = 20 \, x^{3} - 18 \, x}</script></center>
@@ -81,10 +81,10 @@
     sage: test(interacts.calculus.difference_quotient)
     Interactive function <function difference_quotient at ...> with 5 widgets
       title: HTMLText(value=u'<h2>Difference quotient</h2>')
-      f: EvalText(value=u'sin(x)', description=u'f(x)', layout=Layout(max_width=u'81em'))
-      interval: FloatRangeSlider(value=(0.0, 10.0), description=u'Range', max=10.0)
-      a: IntSlider(value=5, description=u'$a$', max=10)
-      x0: IntSlider(value=2, description=u'$x_0$ (start point)', max=10)
+      f: EvalText(value=u'sin(x)', description=u'f(x)')
+      interval: FloatRangeSlider(value=(0.0, 10.0), min=0.0, max=10.0, step=0.1, description=u'Range')
+      a: IntSlider(value=5, min=0, max=10, step=1, description=u'$a$')
+      x0: IntSlider(value=2, min=0, max=10, step=1, description=u'$x_0$ (start point)')
     <h2>Difference Quotient</h2>
     <div style="white-space: normal;">         <a href="https://en.wikipedia.org/wiki/Difference_quotient" target="_blank">         Wikipedia article about difference quotient</a></div>
     <h2>Difference Quotient</h2>
@@ -95,9 +95,9 @@
 
     sage: test(interacts.calculus.quadratic_equation)
     Interactive function <function quadratic_equation at ...> with 3 widgets
-      A: IntSlider(value=1, description=u'A', max=7, min=-7)
-      B: IntSlider(value=1, description=u'B', max=7, min=-7)
-      C: IntSlider(value=-2, description=u'C', max=7, min=-7)
+      A: IntSlider(value=1, min=-7, max=7, step=1, description=u'A')
+      B: IntSlider(value=1, min=-7, max=7, step=1, description=u'B')
+      C: IntSlider(value=-2, min=-7, max=7, step=1, description=u'C')
     <h2>The Solutions of the Quadratic Equation</h2>
     <script type="math/tex">x^2 + x - 2 = 0</script>
     <script type="math/tex">Ax^2 + Bx + C = 0</script>
@@ -106,10 +106,10 @@
     sage: test(interacts.calculus.secant_method)
     Interactive function <function secant_method at ...> with 5 widgets
       title: HTMLText(value=u'<h2>Secant method for numerical root finding</h2>')
-      f: EvalText(value=u'x^2-2', description=u'f(x)', layout=Layout(max_width=u'81em'))
-      interval: IntRangeSlider(value=(0, 4), description=u'range', max=5, min=-5)
-      d: IntSlider(value=3, description=u'10^-d precision', max=16, min=1)
-      maxn: IntSlider(value=10, description=u'max iterations', max=15)
+      f: EvalText(value=u'x^2-2', description=u'f(x)')
+      interval: IntRangeSlider(value=(0, 4), min=-5, max=5, step=1, description=u'range')
+      d: IntSlider(value=3, min=1, max=16, step=1, description=u'10^-d precision')
+      maxn: IntSlider(value=10, min=0, max=15, step=1, description=u'max iterations')
     <script type="math/tex">\text{Precision }h = 10^{-d}=10^{-3}=0.00100</script>
     <script type="math/tex">{c = }1.4144038097709382</script>
     <script type="math/tex">{f(c) = }0.0005381370945443109</script>
@@ -118,11 +118,11 @@
     sage: test(interacts.calculus.newton_method)
     Interactive function <function newton_method at ...> with 7 widgets
       title: HTMLText(value=u'<h2>Newton method</h2>')
-      f: EvalText(value=u'x^2 - 2', description=u'f', layout=Layout(max_width=u'81em'))
-      c: IntSlider(value=6, description=u'Start ($x$)', max=10, min=-10)
-      d: IntSlider(value=3, description=u'$10^{-d}$ precision', max=16, min=1)
-      maxn: IntSlider(value=10, description=u'max iterations', max=15)
-      interval: IntRangeSlider(value=(0, 6), description=u'Interval', max=10, min=-10)
+      f: EvalText(value=u'x^2 - 2', description=u'f')
+      c: IntSlider(value=6, min=-10, max=10, step=1, description=u'Start ($x$)')
+      d: IntSlider(value=3, min=1, max=16, step=1, description=u'$10^{-d}$ precision')
+      maxn: IntSlider(value=10, min=0, max=15, step=1, description=u'max iterations')
+      interval: IntRangeSlider(value=(0, 6), min=-10, max=10, step=1, description=u'Interval')
       list_steps: Checkbox(value=False, description=u'List steps')
     <script type="math/tex">\text{Precision } 2h = 0.001</script>
     <script type="math/tex">{c = }1.4142141576301823</script>
@@ -132,12 +132,12 @@
     sage: test(interacts.calculus.trapezoid_integration)
     Interactive function <function trapezoid_integration at ...> with 7 widgets
       title: HTMLText(value=u'<h2>Trapezoid integration</h2>')
-      f: EvalText(value=u'x^2-5*x + 10', description=u'$f(x)=$', layout=Layout(max_width=u'81em'))
-      n: IntSlider(value=5, description=u'# divisions', min=1)
-      interval_input: ToggleButtons(description=u'Integration interval', options=('from slider', 'from keyboard'), value='from slider')
-      interval_s: IntRangeSlider(value=(0, 8), description=u'slider: ', max=10, min=-10)
-      interval_g: Grid(value=[[0, 8]], children=(Label(value=u'keyboard: '), VBox(children=(EvalText(value=u'0', layout=Layout(max_width=u'5em')),)), VBox(children=(EvalText(value=u'8', layout=Layout(max_width=u'5em')),))))
-      output_form: ToggleButtons(description=u'Computations form', options=('traditional', 'table', 'none'), value='traditional')
+      f: EvalText(value=u'x^2-5*x + 10', description=u'$f(x)=$')
+      n: IntSlider(value=5, min=1, max=100, step=1, description=u'# divisions')
+      interval_input: ToggleButtons(value='from slider', options=['from slider', 'from keyboard'], description=u'Integration interval')
+      interval_s: IntRangeSlider(value=(0, 8), min=-10, max=10, step=1, description=u'slider: ')
+      interval_g: Grid(value=[[0, 8]], description=u'keyboard: ', children=(Label(value=u'keyboard: '), VBox(children=(EvalText(value=u'0'),)), VBox(children=(EvalText(value=u'8'),))))
+      output_form: ToggleButtons(value='traditional', options=['traditional', 'table', 'none'], description=u'Computations form')
     Function <script type="math/tex">f(x)=x^{2} - 5 \, x + 10</script>
     Integral value to seven decimal places is: <script type="math/tex">\displaystyle\int_{0.00}^{8.00} {f(x) \, \mathrm{d}x} = 90.666667</script>
     <BLANKLINE>
@@ -155,12 +155,12 @@
     sage: test(interacts.calculus.simpson_integration)
     Interactive function <function simpson_integration at ...> with 7 widgets
       title: HTMLText(value=u'<h2>Simpson integration</h2>')
-      f: EvalText(value=u'x*sin(x)+x+1', description=u'$f(x)=$', layout=Layout(max_width=u'81em'))
-      n: IntSlider(value=6, description=u'# divisions', min=2, step=2)
-      interval_input: ToggleButtons(description=u'Integration interval', options=('from slider', 'from keyboard'), value='from slider')
-      interval_s: IntRangeSlider(value=(0, 10), description=u'slider: ', max=10, min=-10)
-      interval_g: Grid(value=[[0, 10]], children=(Label(value=u'keyboard: '), VBox(children=(EvalText(value=u'0', layout=Layout(max_width=u'5em')),)), VBox(children=(EvalText(value=u'10', layout=Layout(max_width=u'5em')),))))
-      output_form: ToggleButtons(description=u'Computations form', options=('traditional', 'table', 'none'), value='traditional')
+      f: EvalText(value=u'x*sin(x)+x+1', description=u'$f(x)=$')
+      n: IntSlider(value=6, min=2, max=100, step=2, description=u'# divisions')
+      interval_input: ToggleButtons(value='from slider', options=['from slider', 'from keyboard'], description=u'Integration interval')
+      interval_s: IntRangeSlider(value=(0, 10), min=-10, max=10, step=1, description=u'slider: ')
+      interval_g: Grid(value=[[0, 10]], description=u'keyboard: ', children=(Label(value=u'keyboard: '), VBox(children=(EvalText(value=u'0'),)), VBox(children=(EvalText(value=u'10'),))))
+      output_form: ToggleButtons(value='traditional', options=['traditional', 'table', 'none'], description=u'Computations form')
     Function <script type="math/tex">f(x)=x \sin\left(x\right) + x + 1</script>
     Integral value to seven decimal places is: <script type="math/tex">\displaystyle\int_{0.00}^{10.00} {f(x) \, \mathrm{d}x} = 67.846694</script>
     <BLANKLINE>
@@ -178,10 +178,10 @@
     sage: test(interacts.calculus.bisection_method)
     Interactive function <function bisection_method at ...> with 5 widgets
       title: HTMLText(value=u'<h2>Bisection method</h2>')
-      f: EvalText(value=u'x^2-2', description=u'f(x)', layout=Layout(max_width=u'81em'))
-      interval: IntRangeSlider(value=(0, 4), description=u'range', max=5, min=-5)
-      d: IntSlider(value=3, description=u'$10^{-d}$ precision', max=8, min=1)
-      maxn: IntSlider(value=10, description=u'max iterations', max=50)
+      f: EvalText(value=u'x^2-2', description=u'f(x)')
+      interval: IntRangeSlider(value=(0, 4), min=-5, max=5, step=1, description=u'range')
+      d: IntSlider(value=3, min=1, max=8, step=1, description=u'$10^{-d}$ precision')
+      maxn: IntSlider(value=10, min=0, max=50, step=1, description=u'max iterations')
     <script type="math/tex">\text{Precision }h = 10^{-d}=10^{-3}=0.00100</script>
     <script type="math/tex">{c = }1.4140625</script>
     <script type="math/tex">{f(c) = }-0.00042724609375</script>
@@ -190,12 +190,12 @@
     sage: test(interacts.calculus.riemann_sum)
     Manual interactive function <function riemann_sum at ...> with 9 widgets
       title: HTMLText(value=u'<h2>Riemann integral with random sampling</h2>')
-      f: EvalText(value=u'x^2+1', description=u'$f(x)=$', layout=Layout(max_width=u'41em'))
-      n: IntSlider(value=5, description=u'# divisions', max=30, min=1)
+      f: EvalText(value=u'x^2+1', description=u'$f(x)=$')
+      n: IntSlider(value=5, min=1, max=30, step=1, description=u'# divisions')
       hr1: HTMLText(value=u'<hr>')
-      interval_input: ToggleButtons(description=u'Integration interval', options=('from slider', 'from keyboard'), value='from slider')
-      interval_s: IntRangeSlider(value=(0, 2), description=u'slider: ', max=10, min=-5)
-      interval_g: Grid(value=[[0, 2]], children=(Label(value=u'keyboard: '), VBox(children=(EvalText(value=u'0', layout=Layout(max_width=u'5em')),)), VBox(children=(EvalText(value=u'2', layout=Layout(max_width=u'5em')),))))
+      interval_input: ToggleButtons(value='from slider', options=['from slider', 'from keyboard'], description=u'Integration interval')
+      interval_s: IntRangeSlider(value=(0, 2), min=-5, max=10, step=1, description=u'slider: ')
+      interval_g: Grid(value=[[0, 2]], description=u'keyboard: ', children=(Label(value=u'keyboard: '), VBox(children=(EvalText(value=u'0'),)), VBox(children=(EvalText(value=u'2'),))))
       hr2: HTMLText(value=u'<hr>')
       list_table: Checkbox(value=False, description=u'List table')
     <small>Adjust your data and click Update button. Click repeatedly for another random values.</small>
@@ -208,10 +208,10 @@
     Interactive function <function function_tool at ...> with 7 widgets
       f: EvalText(value=u'sin(x)', description=u'f')
       g: EvalText(value=u'cos(x)', description=u'g')
-      xrange: IntRangeSlider(value=(0, 1), description=u'x-range', max=3, min=-3)
+      xrange: IntRangeSlider(value=(0, 1), min=-3, max=3, step=1, description=u'x-range')
       yrange: Text(value=u'auto', description=u'yrange')
-      a: IntSlider(value=1, description=u'a', max=3, min=-1)
-      action: ToggleButtons(description=u'h = ', options=('f', 'df/dx', 'int f', 'num f', 'den f', '1/f', 'finv', 'f+a', 'f-a', 'f*a', 'f/a', 'f^a', 'f(x+a)', 'f(x*a)', 'f+g', 'f-g', 'f*g', 'f/g', 'f(g)'), value='f')
+      a: IntSlider(value=1, min=-1, max=3, step=1, description=u'a')
+      action: ToggleButtons(value='f', options=['f', 'df/dx', 'int f', 'num f', 'den f', '1/f', 'finv', 'f+a', 'f-a', 'f*a', 'f/a', 'f^a', 'f(x+a)', 'f(x*a)', 'f+g', 'f-g', 'f*g', 'f/g', 'f(g)'], description=u'h = ')
       do_plot: Checkbox(value=True, description=u'Draw Plots')
     <center><font color="red"><script type="math/tex">f = \sin\left(x\right)</script></font></center>
     <center><font color="green"><script type="math/tex">g = \cos\left(x\right)</script></font></center>
@@ -219,46 +219,46 @@
 
     sage: test(interacts.fractals.mandelbrot)
     Interactive function <function mandelbrot at ...> with 6 widgets
-      expo: FloatSlider(value=2.0, description=u'expo', max=10.0, min=-10.0)
-      iterations: IntSlider(value=20, description=u'# iterations', min=1)
-      zoom_x: FloatRangeSlider(value=(-2.0, 1.0), description=u'Zoom X', max=2.0, min=-2.0, step=0.01)
-      zoom_y: FloatRangeSlider(value=(-1.5, 1.5), description=u'Zoom Y', max=2.0, min=-2.0, step=0.01)
-      plot_points: IntSlider(value=150, description=u'plot points', max=400, min=20, step=20)
-      dpi: IntSlider(value=80, description=u'dpi', max=200, min=20, step=10)
+      expo: FloatSlider(value=2.0, min=-10.0, max=10.0, step=0.1, description=u'expo')
+      iterations: IntSlider(value=20, min=1, max=100, step=1, description=u'# iterations')
+      zoom_x: FloatRangeSlider(value=(-2.0, 1.0), min=-2.0, max=2.0, step=0.01, description=u'Zoom X')
+      zoom_y: FloatRangeSlider(value=(-1.5, 1.5), min=-2.0, max=2.0, step=0.01, description=u'Zoom Y')
+      plot_points: IntSlider(value=150, min=20, max=400, step=20, description=u'plot points')
+      dpi: IntSlider(value=80, min=20, max=200, step=10, description=u'dpi')
     <h2>Mandelbrot Fractal</h2>
     Recursive Formula: <script type="math/tex">z \leftarrow z^{2.00} + c</script> for <script type="math/tex">c \in \mathbb{C}</script>
 
     sage: test(interacts.fractals.julia)
     Interactive function <function julia at ...> with 8 widgets
-      expo: FloatSlider(value=2.0, description=u'expo', max=10.0, min=-10.0)
-      c_real: FloatSlider(value=0.5, description=u'real part const.', max=2.0, min=-2.0, step=0.01)
-      c_imag: FloatSlider(value=0.5, description=u'imag part const.', max=2.0, min=-2.0, step=0.01)
-      iterations: IntSlider(value=20, description=u'# iterations', min=1)
-      zoom_x: FloatRangeSlider(value=(-1.5, 1.5), description=u'Zoom X', max=2.0, min=-2.0, step=0.01)
-      zoom_y: FloatRangeSlider(value=(-1.5, 1.5), description=u'Zoom Y', max=2.0, min=-2.0, step=0.01)
-      plot_points: IntSlider(value=150, description=u'plot points', max=400, min=20, step=20)
-      dpi: IntSlider(value=80, description=u'dpi', max=200, min=20, step=10)
+      expo: FloatSlider(value=2.0, min=-10.0, max=10.0, step=0.1, description=u'expo')
+      c_real: FloatSlider(value=0.5, min=-2.0, max=2.0, step=0.01, description=u'real part const.')
+      c_imag: FloatSlider(value=0.5, min=-2.0, max=2.0, step=0.01, description=u'imag part const.')
+      iterations: IntSlider(value=20, min=1, max=100, step=1, description=u'# iterations')
+      zoom_x: FloatRangeSlider(value=(-1.5, 1.5), min=-2.0, max=2.0, step=0.01, description=u'Zoom X')
+      zoom_y: FloatRangeSlider(value=(-1.5, 1.5), min=-2.0, max=2.0, step=0.01, description=u'Zoom Y')
+      plot_points: IntSlider(value=150, min=20, max=400, step=20, description=u'plot points')
+      dpi: IntSlider(value=80, min=20, max=200, step=10, description=u'dpi')
     <h2>Julia Fractal</h2>
     Recursive Formula: <script type="math/tex">z \leftarrow z^{2.00} + (0.50+0.50*\mathbb{I})</script>
 
     sage: test(interacts.fractals.cellular_automaton)
     Interactive function <function cellular_automaton at ...> with 3 widgets
-      N: IntSlider(value=100, description=u'Number of iterations', max=500, min=1)
-      rule_number: IntSlider(value=110, description=u'Rule number', max=255)
-      size: IntSlider(value=6, description=u'size of graphic', max=11, min=1)
+      N: IntSlider(value=100, min=1, max=500, step=1, description=u'Number of iterations')
+      rule_number: IntSlider(value=110, min=0, max=255, step=1, description=u'Rule number')
+      size: IntSlider(value=6, min=1, max=11, step=1, description=u'size of graphic')
     <h2>Cellular Automaton</h2><div style="white-space: normal;">"A cellular automaton is a collection of "colored" cells          on a grid of specified shape that evolves through a number of          discrete time steps according to a set of rules based on the          states of neighboring cells." &mdash;          <a target="_blank" href="http://mathworld.wolfram.com/CellularAutomaton.html">Mathworld,         Cellular Automaton</a></div>         <div>Rule 110 expands to 01110110</div>
 
     sage: test(interacts.geometry.unit_circle)
     Interactive function <function unit_circle at ...> with 2 widgets
-      function: Dropdown(description=u'function', options=(('sin(x)', 0), ('cos(x)', 1), ('tan(x)', 2)), value=0)
-      x: TransformFloatSlider(value=0.0, description=u'x', max=6.283185307179586, step=0.015707963267948967)
+      function: Dropdown(value=0, options=[('sin(x)', 0), ('cos(x)', 1), ('tan(x)', 2)], description=u'function')
+      x: TransformFloatSlider(value=0.0, min=0.0, max=6.283185307179586, step=0.015707963267948967, description=u'x')
     <div style="white-space: normal;">Lines of the same color have         the same length</div>
 
     sage: test(interacts.geometry.trigonometric_properties_triangle)
     Interactive function <function trigonometric_properties_triangle at ...> with 3 widgets
-      a0: IntSlider(value=30, description=u'A', max=360)
-      a1: IntSlider(value=180, description=u'B', max=360)
-      a2: IntSlider(value=300, description=u'C', max=360)
+      a0: IntSlider(value=30, min=0, max=360, step=1, description=u'A')
+      a1: IntSlider(value=180, min=0, max=360, step=1, description=u'B')
+      a2: IntSlider(value=300, min=0, max=360, step=1, description=u'C')
     <h2>Trigonometric Properties of a Triangle</h2>
     <script type="math/tex">\angle A = {60.000}^{\circ},</script> <script type="math/tex">\angle B = {45.000}^{\circ},</script> <script type="math/tex">\angle C = {75.000}^{\circ}</script>
     <script type="math/tex">AB = 1.931852</script>, <script type="math/tex">BC = 1.732051</script>, <script type="math/tex">CA = 1.414214</script>
@@ -267,9 +267,9 @@
     sage: test(interacts.geometry.special_points)
     Interactive function <function special_points at ...> with 10 widgets
       title: HTMLText(value=u'<h2>Special points in triangle</h2>')
-      a0: IntSlider(value=30, description=u'A', max=360)
-      a1: IntSlider(value=180, description=u'B', max=360)
-      a2: IntSlider(value=300, description=u'C', max=360)
+      a0: IntSlider(value=30, min=0, max=360, step=1, description=u'A')
+      a1: IntSlider(value=180, min=0, max=360, step=1, description=u'B')
+      a2: IntSlider(value=300, min=0, max=360, step=1, description=u'C')
       show_median: Checkbox(value=False, description=u'Medians')
       show_pb: Checkbox(value=False, description=u'Perpendicular Bisectors')
       show_alt: Checkbox(value=False, description=u'Altitudes')
--- a/sage/src/sage/repl/ipython_kernel/interact.py
+++ b/sage/src/sage/repl/ipython_kernel/interact.py
@@ -9,7 +9,7 @@
 
 We need to setup a proper test environment for widgets::
 
-    sage: from ipywidgets.widgets.tests.utils import setup_test_comm
+    sage: from ipywidgets.widgets.tests import setup_test_comm
     sage: setup_test_comm()
 
 EXAMPLES::
@@ -18,10 +18,10 @@
     sage: @interact
     ....: def f(x=(0,10)):
     ....:     pass
-    Interactive function <function f at ...> with 1 widget
-      x: IntSlider(value=5, description='x', max=10)
+    Interactive function <function f ...> with 1 widget
+      x: IntSlider(value=5, min=0, max=10, step=1, description='x')
     sage: f.widget.children
-    (IntSlider(value=5, description='x', max=10), Output())
+    (IntSlider(value=5, min=0, max=10, step=1, description='x'), Output())
 """
 
 # ****************************************************************************
@@ -56,10 +56,10 @@
         sage: from sage.repl.ipython_kernel.interact import sage_interactive
         sage: def myfunc(x=10, y="hello", z=None): pass
         sage: sage_interactive(myfunc, x=(0,100), z=["one", "two", "three"])
-        Interactive function <function myfunc at ...> with 3 widgets
-          x: IntSlider(value=10, description='x')
+        Interactive function <function myfunc ...> with 3 widgets
+          x: IntSlider(value=10, min=0, max=100, step=1, description='x')
           y: Text(value='hello', description='y')
-          z: Dropdown(description='z', options=('one', 'two', 'three'), value=None)
+          z: Dropdown(value='one', options=['one', 'two', 'three'], description='z')
     """
     def __init__(*args, **kwds):
         """
@@ -196,15 +196,15 @@
 
             sage: from sage.repl.ipython_kernel.interact import sage_interactive
             sage: sage_interactive.widget_from_tuple( (0, 10) )
-            IntSlider(value=5, max=10)
+            IntSlider(value=5, min=0, max=10, step=1)
             sage: sage_interactive.widget_from_tuple( ("number", (0, 10)) )
-            IntSlider(value=5, description='number', max=10)
+            IntSlider(value=5, min=0, max=10, step=1, description='number')
             sage: sage_interactive.widget_from_tuple( (3, (0, 10)) )
-            IntSlider(value=3, max=10)
+            IntSlider(value=3, min=0, max=10, step=1)
             sage: sage_interactive.widget_from_tuple((2, dict(one=1, two=2, three=3)))
-            Dropdown(index=1, options={'one': 1, 'two': 2, 'three': 3}, value=2)
+            Dropdown(value=2, options={'one': 1, 'two': 2, 'three': 3})
             sage: sage_interactive.widget_from_tuple( (sqrt(2), pi) )
-            FloatSlider(value=2.277903107981444, max=3.141592653589793, min=1.4142135623730951)
+            FloatSlider(value=2.277903107981444, min=1.4142135623730951, max=3.141592653589793, step=0.1)
 
         TESTS:
 
@@ -246,17 +246,17 @@
 
             sage: from sage.repl.ipython_kernel.interact import sage_interactive
             sage: sage_interactive.widget_from_iterable([1..5])
-            Dropdown(options=(1, 2, 3, 4, 5), value=1)
+            Dropdown(value=1, options=[1, 2, 3, 4, 5])
             sage: sage_interactive.widget_from_iterable(iter([1..5]))
-            SelectionSlider(options=(1, 2, 3, 4, 5), value=1)
+            SelectionSlider(value=1, options=[1, 2, 3, 4, 5])
             sage: sage_interactive.widget_from_iterable((1..5))
-            SelectionSlider(options=(1, 2, 3, 4, 5), value=1)
+            SelectionSlider(value=1, options=[1, 2, 3, 4, 5])
             sage: sage_interactive.widget_from_iterable(x for x in [1..5])
-            SelectionSlider(options=(1, 2, 3, 4, 5), value=1)
+            SelectionSlider(value=1, options=[1, 2, 3, 4, 5])
             sage: def gen():
             ....:     yield 1; yield 2; yield 3; yield 4; yield 5
             sage: sage_interactive.widget_from_iterable(gen())
-            SelectionSlider(options=(1, 2, 3, 4, 5), value=1)
+            SelectionSlider(value=1, options=[1, 2, 3, 4, 5])
         """
         if isinstance(abbrev, Iterator):
             return SelectionSlider(options=list(abbrev))
--- a/sage/src/sage/repl/ipython_kernel/widgets.py
+++ b/sage/src/sage/repl/ipython_kernel/widgets.py
@@ -8,7 +8,7 @@
 
 We need to setup a proper test environment for widgets::
 
-    sage: from ipywidgets.widgets.tests.utils import setup_test_comm
+    sage: from ipywidgets.widgets.tests import setup_test_comm
     sage: setup_test_comm()
 """
 
@@ -131,7 +131,7 @@
         sage: class TransformToggleButtons(TransformWidget, ToggleButtons): pass
         sage: w = TransformToggleButtons(options=["pi", "e"], transform=lambda x: x+x)
         sage: w
-        TransformToggleButtons(options=('pi', 'e'), value='pi')
+        TransformToggleButtons(value='pi', options=['pi', 'e'])
         sage: w.get_interact_value()
         'pipi'
     """
@@ -201,7 +201,7 @@
         sage: class EvalToggleButtons(EvalWidget, ToggleButtons): pass
         sage: w = EvalToggleButtons(options=["pi", "e"], transform=lambda x: x+x)
         sage: w
-        EvalToggleButtons(options=('pi', 'e'), value='pi')
+        EvalToggleButtons(value='pi', options=['pi', 'e'])
         sage: w.get_interact_value()
         2*pi
     """
@@ -216,7 +216,7 @@
             sage: class EvalDropdown(EvalWidget, Dropdown): pass
             sage: w = EvalDropdown(options=["the_answer"], transform=RR)
             sage: w
-            EvalDropdown(options=('the_answer',), value='the_answer')
+            EvalDropdown(value='the_answer', options=['the_answer'])
             sage: the_answer = 42
             sage: w.get_value()
             42
@@ -236,7 +236,7 @@
         sage: from sage.repl.ipython_kernel.widgets import TransformIntSlider
         sage: w = TransformIntSlider(min=0, max=100, value=7, transform=lambda x: x^2)
         sage: w
-        TransformIntSlider(value=7)
+        TransformIntSlider(value=7, min=0, max=100, step=1)
         sage: w.get_interact_value()
         49
     """
@@ -253,7 +253,7 @@
         sage: from sage.repl.ipython_kernel.widgets import TransformFloatSlider
         sage: w = TransformFloatSlider(min=0, max=100, value=7, transform=lambda x: sqrt(x))
         sage: w
-        TransformFloatSlider(value=7.0)
+        TransformFloatSlider(value=7.0, min=0.0, max=100.0, step=0.1)
         sage: w.get_interact_value()
         2.6457513110645907
     """
@@ -270,7 +270,7 @@
         sage: from sage.repl.ipython_kernel.widgets import TransformIntRangeSlider
         sage: w = TransformIntRangeSlider(min=0, max=100, value=(7,9), transform=lambda x: x[1]-x[0])
         sage: w
-        TransformIntRangeSlider(value=(7, 9))
+        TransformIntRangeSlider(value=(7, 9), min=0, max=100, step=1)
         sage: w.get_interact_value()
         2
     """
@@ -287,7 +287,7 @@
         sage: from sage.repl.ipython_kernel.widgets import TransformFloatRangeSlider
         sage: w = TransformFloatRangeSlider(min=0, max=100, value=(7,9), transform=lambda x: x[1]-x[0])
         sage: w
-        TransformFloatRangeSlider(value=(7.0, 9.0))
+        TransformFloatRangeSlider(value=(7.0, 9.0), min=0.0, max=100.0, step=1.0)
         sage: w.get_interact_value()
         2.0
     """
@@ -364,7 +364,7 @@
 
 class SageColorPicker(ColorPicker):
     """
-    A color picker widget returning a Sage :class:`Color`.
+    A color picker widget return a Sage :class:`Color`.
 
     EXAMPLES::
 
@@ -427,7 +427,7 @@
             sage: w = Grid(2, 2, lambda i,j: EvalText(str(j+4*i)),
             ....:         description="2x2 matrix", transform=matrix)
             sage: w
-            Grid(value=[[0, 1], [4, 5]], children=(Label(value='2x2 matrix'), VBox(children=(EvalText(value='0'), EvalText(value='4'))), VBox(children=(EvalText(value='1'), EvalText(value='5')))))
+            Grid(value=[[0, 1], [4, 5]], description='2x2 matrix', children=(Label(value='2x2 matrix'), VBox(children=(EvalText(value='0'), EvalText(value='4'))), VBox(children=(EvalText(value='1'), EvalText(value='5')))))
             sage: w.get_interact_value()
             [0 1]
             [4 5]
--- a/sage/src/sage/repl/ipython_kernel/widgets_sagenb.py
+++ b/sage/src/sage/repl/ipython_kernel/widgets_sagenb.py
@@ -8,7 +8,7 @@
 
 We need to setup a proper test environment for widgets::
 
-    sage: from ipywidgets.widgets.tests.utils import setup_test_comm
+    sage: from ipywidgets.widgets.tests import setup_test_comm
     sage: setup_test_comm()
 
 EXAMPLES::
@@ -78,7 +78,7 @@
 
         sage: w = input_box("4+5", type=str, label="enter a string")
         sage: w
-        TransformText(value='4+5', description='enter a string', layout=Layout(max_width='81em'))
+        TransformText(value='4+5', description='enter a string')
         sage: w.get_interact_value()
         '4+5'
 
@@ -86,7 +86,7 @@
 
         sage: w = input_box("4+5")
         sage: w
-        EvalText(value='4+5', layout=Layout(max_width='81em'))
+        EvalText(value='4+5')
         sage: w.get_interact_value()
         9
 
@@ -95,7 +95,7 @@
 
         sage: w = input_box("4+5", type=float)
         sage: w
-        EvalText(value='4+5', layout=Layout(max_width='81em'))
+        EvalText(value='4+5')
         sage: w.get_interact_value()
         9.0
 
@@ -103,7 +103,7 @@
 
         sage: w = input_box("4+5", type=sqrt)
         sage: w
-        EvalText(value='4+5', layout=Layout(max_width='81em'))
+        EvalText(value='4+5')
         sage: w.get_interact_value()
         3
 
@@ -111,10 +111,10 @@
 
         sage: w = input_box("4+5", width=100, height=1)
         sage: w
-        EvalText(value='4+5', layout=Layout(max_width='101em'))
+        EvalText(value='4+5')
         sage: w = input_box("4+5", width=100, height=5)
         sage: w
-        EvalTextarea(value='4+5', layout=Layout(max_width='101em'))
+        EvalTextarea(value='4+5')
 
     TESTS::
 
@@ -179,37 +179,37 @@
 
         sage: from sage.repl.ipython_kernel.all_jupyter import slider
         sage: slider(5, label="slide me")
-        TransformIntSlider(value=5, description='slide me', min=5)
+        TransformIntSlider(value=5, min=5, max=100, step=1, description='slide me')
         sage: slider(5, 20)
-        TransformIntSlider(value=5, max=20, min=5)
+        TransformIntSlider(value=5, min=5, max=20, step=1)
         sage: slider(5, 20, 0.5)
-        TransformFloatSlider(value=5.0, max=20.0, min=5.0, step=0.5)
+        TransformFloatSlider(value=5.0, min=5.0, max=20.0, step=0.5)
         sage: slider(5, 20, default=12)
-        TransformIntSlider(value=12, max=20, min=5)
+        TransformIntSlider(value=12, min=5, max=20, step=1)
 
     The parent of the inputs determines the parent of the value::
 
         sage: w = slider(5); w
-        TransformIntSlider(value=5, min=5)
+        TransformIntSlider(value=5, min=5, max=100, step=1)
         sage: parent(w.get_interact_value())
         Integer Ring
         sage: w = slider(int(5)); w
-        IntSlider(value=5, min=5)
+        IntSlider(value=5, min=5, max=100, step=1)
         sage: parent(w.get_interact_value())
         <... 'int'>
         sage: w = slider(5, 20, step_size=RDF("0.1")); w
-        TransformFloatSlider(value=5.0, max=20.0, min=5.0)
+        TransformFloatSlider(value=5.0, min=5.0, max=20.0, step=0.1)
         sage: parent(w.get_interact_value())
         Real Double Field
         sage: w = slider(5, 20, step_size=10/3); w
-        SelectionSlider(index=2, options=(5, 25/3, 35/3, 15, 55/3), value=35/3)
+        SelectionSlider(value=35/3, options=[5, 25/3, 35/3, 15, 55/3])
         sage: parent(w.get_interact_value())
         Rational Field
 
     Symbolic input is evaluated numerically::
 
         sage: w = slider(e, pi); w
-        TransformFloatSlider(value=2.718281828459045, max=3.141592653589793, min=2.718281828459045)
+        TransformFloatSlider(value=2.718281828459045, min=2.718281828459045, max=3.141592653589793, step=0.1)
         sage: parent(w.get_interact_value())
         Real Field with 53 bits of precision
 
@@ -217,7 +217,7 @@
     possible values::
 
         sage: slider(range(10), default=17/10)
-        SelectionSlider(index=2, options=(0, 1, 2, 3, 4, 5, 6, 7, 8, 9), value=2)
+        SelectionSlider(value=2, options=[0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
 
     TESTS::
 
@@ -362,26 +362,26 @@
 
         sage: from sage.repl.ipython_kernel.all_jupyter import range_slider
         sage: range_slider(5, label="slide me")
-        TransformIntRangeSlider(value=(28, 76), description='slide me', min=5)
+        TransformIntRangeSlider(value=(28, 76), min=5, max=100, step=1, description='slide me')
         sage: range_slider(5, 20)
-        TransformIntRangeSlider(value=(8, 16), max=20, min=5)
+        TransformIntRangeSlider(value=(8, 16), min=5, max=20, step=1)
         sage: range_slider(5, 20, 0.5)
-        TransformFloatRangeSlider(value=(8.75, 16.25), max=20.0, min=5.0, step=0.5)
+        TransformFloatRangeSlider(value=(8.75, 16.25), min=5.0, max=20.0, step=0.5)
         sage: range_slider(5, 20, default=(12,15))
-        TransformIntRangeSlider(value=(12, 15), max=20, min=5)
+        TransformIntRangeSlider(value=(12, 15), min=5, max=20, step=1)
 
     The parent of the inputs determines the parent of the value::
 
         sage: w = range_slider(5); w
-        TransformIntRangeSlider(value=(28, 76), min=5)
+        TransformIntRangeSlider(value=(28, 76), min=5, max=100, step=1)
         sage: [parent(x) for x in w.get_interact_value()]
         [Integer Ring, Integer Ring]
         sage: w = range_slider(int(5)); w
-        IntRangeSlider(value=(28, 76), min=5)
+        IntRangeSlider(value=(28, 76), min=5, max=100, step=1)
         sage: [parent(x) for x in w.get_interact_value()]
         [<... 'int'>, <... 'int'>]
         sage: w = range_slider(5, 20, step_size=RDF("0.1")); w
-        TransformFloatRangeSlider(value=(8.75, 16.25), max=20.0, min=5.0)
+        TransformFloatRangeSlider(value=(8.75, 16.25), min=5.0, max=20.0, step=0.1)
         sage: [parent(x) for x in w.get_interact_value()]
         [Real Double Field, Real Double Field]
 
@@ -450,31 +450,31 @@
 
         sage: from sage.repl.ipython_kernel.all_jupyter import selector
         sage: selector(range(5), label="choose one")
-        Dropdown(description='choose one', options=(0, 1, 2, 3, 4), value=0)
+        Dropdown(value=0, options=[0, 1, 2, 3, 4], description='choose one')
         sage: selector(range(5), buttons=True, default=4)
-        ToggleButtons(index=4, options=(0, 1, 2, 3, 4), value=4)
+        ToggleButtons(value=4, options=[0, 1, 2, 3, 4])
 
     Apart from a simple list, ``values`` can be given as a list of
     2-tuples ``(value, label)``::
 
         sage: selector([(1,"one"), (2,"two"), (3,"three")])
-        Dropdown(options=(('one', 1), ('two', 2), ('three', 3)), value=1)
+        Dropdown(value=1, options=[('one', 1), ('two', 2), ('three', 3)])
         sage: selector([(1,"one"), (2,"two"), (3,"three")], buttons=True)
-        ToggleButtons(options=(('one', 1), ('two', 2), ('three', 3)), value=1)
+        ToggleButtons(value=1, options=[('one', 1), ('two', 2), ('three', 3)])
 
     A dict of ``label:value`` pairs is also allowed. Since a ``dict``
     is not ordered, it is better to use an :class:`OrderedDict`::
 
         sage: from collections import OrderedDict
         sage: selector(OrderedDict(one=1, two=2, three=3))
-        Dropdown(options=OrderedDict([('one', 1), ('two', 2), ('three', 3)]), value=1)
+        Dropdown(value=1, options=OrderedDict([('one', 1), ('two', 2), ('three', 3)]))
         sage: selector(OrderedDict(one=1, two=2, three=3), buttons=True)
-        ToggleButtons(options=OrderedDict([('one', 1), ('two', 2), ('three', 3)]), value=1)
+        ToggleButtons(value=1, options=OrderedDict([('one', 1), ('two', 2), ('three', 3)]))
 
     The values can be any kind of object:
 
         sage: selector([sin(x^2), GF(29), EllipticCurve('37a1')])
-        Dropdown(options=(sin(x^2), Finite Field of size 29, Elliptic Curve defined by y^2 + y = x^3 - x over Rational Field), value=sin(x^2))
+        Dropdown(value=sin(x^2), options=(sin(x^2), Finite Field of size 29, Elliptic Curve defined by y^2 + y = x^3 - x over Rational Field))
     """
     if isinstance(values, Sequence):
         values = list(values)
@@ -527,14 +527,14 @@
 
         sage: from sage.repl.ipython_kernel.all_jupyter import input_grid
         sage: input_grid(2, 2, default=42, label="answers")
-        Grid(value=[[42, 42], [42, 42]], children=(Label(value='answers'), VBox(children=(EvalText(value='42', layout=Layout(max_width='5em')), EvalText(value='42', layout=Layout(max_width='5em')))), VBox(children=(EvalText(value='42', layout=Layout(max_width='5em')), EvalText(value='42', layout=Layout(max_width='5em'))))))
+        Grid(value=[[42, 42], [42, 42]], description='answers', children=(Label(value='answers'), VBox(children=(EvalText(value='42'), EvalText(value='42'))), VBox(children=(EvalText(value='42'), EvalText(value='42')))))
         sage: w = input_grid(2, 2, default=[[cos(x), sin(x)], [-sin(x), cos(x)]], to_value=matrix); w
-        Grid(value=[[cos(x), sin(x)], [-sin(x), cos(x)]], children=(Label(value=''), VBox(children=(EvalText(value='cos(x)', layout=Layout(max_width='5em')), EvalText(value='-sin(x)', layout=Layout(max_width='5em')))), VBox(children=(EvalText(value='sin(x)', layout=Layout(max_width='5em')), EvalText(value='cos(x)', layout=Layout(max_width='5em'))))))
+        Grid(value=[[cos(x), sin(x)], [-sin(x), cos(x)]], children=(Label(value=''), VBox(children=(EvalText(value='cos(x)'), EvalText(value='-sin(x)'))), VBox(children=(EvalText(value='sin(x)'), EvalText(value='cos(x)')))))
         sage: w.get_interact_value()
         [ cos(x)  sin(x)]
         [-sin(x)  cos(x)]
         sage: w = input_grid(2, 2, default=[1, x, x^2, x^3], to_value=lambda x: x[1][1]); w
-        Grid(value=[[1, x], [x^2, x^3]], children=(Label(value=''), VBox(children=(EvalText(value='1', layout=Layout(max_width='5em')), EvalText(value='x^2', layout=Layout(max_width='5em')))), VBox(children=(EvalText(value='x', layout=Layout(max_width='5em')), EvalText(value='x^3', layout=Layout(max_width='5em'))))))
+        Grid(value=[[1, x], [x^2, x^3]], children=(Label(value=''), VBox(children=(EvalText(value='1'), EvalText(value='x^2'))), VBox(children=(EvalText(value='x'), EvalText(value='x^3')))))
         sage: w.get_interact_value()
         x^3
     """
--- a/sage/src/sage/cpython/debug.pyx
+++ b/sage/src/sage/cpython/debug.pyx
@@ -111,7 +111,7 @@
           returning log (<class 'sage.interfaces.expect.FunctionElement'>)
         sage: from ipywidgets import IntSlider
         sage: _ = getattr_debug(IntSlider(), "value")
-        getattr_debug(obj=IntSlider(value=0), name='value'):
+        getattr_debug(obj=IntSlider(value=0, min=0, max=100, step=1), name='value'):
           type(obj) = <class 'ipywidgets.widgets.widget_int.IntSlider'>
           object has __dict__ slot (<class 'dict'>)
           found 'value' in dict of <class 'ipywidgets.widgets.widget_int._Int'>
